editor/plugins/animation_player_editor_plugin.cpp

```diff
@@ -1979,14 +2052,89 @@ void AnimationPlayerEditorPlugin::_update_keying() {
 }
 
 void AnimationPlayerEditorPlugin::edit(Object *p_object) {
+	player = nullptr;
 	if (!p_object) {
 		return;
 	}
-	anim_editor->edit(Object::cast_to<AnimationPlayer>(p_object));
+	last_manager = p_object->get_instance_id();
+
+	AnimationManager *src_node = Object::cast_to<AnimationManager>(p_object);
+	bool is_dummy = false;
+	if (!p_object->is_class("AnimationPlayer")) {
+		// If it needs dummy AnimationPlayer, assign original AnimationManager to LibraryEditor.
+		_update_dummy_player(src_node);
+
+		is_dummy = true;
+
+		if (!src_node->is_connected("manager_updated", callable_mp(this, &AnimationPlayerEditorPlugin::_update_dummy_player))) {
+			src_node->connect("manager_updated", callable_mp(this, &AnimationPlayerEditorPlugin::_update_dummy_player).bind(src_node), CONNECT_DEFERRED);
+		}
+
+		// Does AnimationTree have AnimationPlayer?
+		if (p_object->is_class("AnimationTree")) {
+			AnimationTree *src_tree = Object::cast_to<AnimationTree>(p_object);
+			Node *src_player = src_tree->get_node_or_null(src_tree->get_animation_player());
+			if (src_player) {
+				p_object = src_player;
+			}
+		}
+	} else {
+		_clear_dummy_player();
+		player = Object::cast_to<AnimationPlayer>(p_object);
+	}
+
+	anim_editor->edit(src_node, player, Object::cast_to<AnimationManager>(p_object), is_dummy);
+}
+
+void AnimationPlayerEditorPlugin::_clear_dummy_player() {
+	if (!dummy_player) {
+		return;
+	}
+	Node *parent = dummy_player->get_parent();
+	if (parent) {
+		parent->call_deferred("remove_child", dummy_player);
+	}
+	dummy_player = nullptr;
+}
+
+void AnimationPlayerEditorPlugin::_update_dummy_player(AnimationManager *p_manager) {
+	// Check current editing object.
+	if (p_manager->get_instance_id() != last_manager && p_manager->is_connected("manager_updated", callable_mp(this, &AnimationPlayerEditorPlugin::_update_dummy_player))) {
+		p_manager->disconnect("manager_updated", callable_mp(this, &AnimationPlayerEditorPlugin::_update_dummy_player));
+		return;
+	}
+
+	// Add dummy player to scene.
+	if (!dummy_player) {
+		dummy_player = memnew(AnimationPlayer);
+		Node *parent = p_manager->get_parent();
+		if (parent) {
+			parent->add_child(dummy_player);
+		}
+	}
+	player = dummy_player;
+
+	// Convert AnimationTree (AnimationManager) to AnimationPlayer.
+	AnimationManager *default_node = memnew(AnimationManager);
+	List<PropertyInfo> pinfo;
+	default_node->get_property_list(&pinfo);
+	for (const PropertyInfo &E : pinfo) {
+		if (!(E.usage & PROPERTY_USAGE_STORAGE)) {
+			continue;
+		}
+		if (!(E.name == "script" && E.name == "active" && E.name == "predictable" && E.name == "root_motion_track")) {
```

Comment:
I think and chain is faster since it can detect false early.

Reply:
This condition looks backwards, it will always be false, is it not supposed to be:
```suggestion
		if (!(E.name == "script" || E.name == "active" || E.name == "predictable" || E.name == "root_motion_track")) {
```
I.e.
```suggestion
		if (E.name != "script" && E.name != "active" && E.name != "predictable" && E.name != "root_motion_track") {
```